File: Scala/euler/src/re/_1r/euler/problems/70.scala
package re._1r.euler
package problems

import math._
import utils.Primes._
import utils.NumberOps.intOps
import collection.mutable.HashMap

object Euler70 {
  val pfMap = HashMap[Int, Array[Int]]()
  def primeFactors(n: Int, primes: Array[Int]): Array[Int] = {
    if(n % 10000 == 0) println(n)
    if(n==1) Array()
    else if(primes.last == n) Array(n)
    else {
      def divisor = primes.find(n % _ == 0).getOrElse(1)
      divisor +: pfMap(n / divisor)
    }
  }
  def n = 10000000
  val primes = genPrimes(n)
  def result = {
    (2 until n).foreach(i => pfMap += ((i, primeFactors(i, primes.takeWhile(_ <= i)))))
    pfMap.map{case (a, b) => (a, b.distinct.map(_.toDouble).fold(a.toDouble)((acc, v) => acc * (1 - 1 / v.toDouble)).toInt)}.filter{case (a, b) => a.digits.sorted == b.digits.sorted}.minBy{case (a, b) => a.toDouble / b.toDouble}
  }
}

File: Elixir/70.exs
defmodule Euler70 do
  def filterle(list, n) do
    Enum.take_while(list, fn x -> x <= n end)
  end

  def genPrimes(max) do
    k = trunc((max - 2) / 2)
    notPrimes = Enum.reduce(1..k, Enum.to_list(1..max), fn i, primes ->
      Enum.reduce(i..k, primes, fn j, primes ->
        if i + j + 2 * i * j <= k do
          List.delete_at(primes, 2 * (i + j + 2 * i * j) + 1)
        else
          primes
        end
      end)
    end)
    [2 | Enum.take_every(notPrimes, 2)]
  end

  def primeFactors(n, primes, pfMap) do
    Map.put(pfMap, n, cond do
      n == 1 -> []
      List.last(primes) == n -> [n]
      true ->
        divisor = Enum.find(primes, 1, fn x -> rem(n, x) == 0 end)
        [divisor | Map.get(pfMap, div(n, divisor))]
    end)
  end

  def phi(x, pf) do
    trunc(Enum.reduce(Enum.uniq(pf), x, fn y, acc ->
      acc * (1 - 1 / y)
    end))
  end

  def result do
    n = 100000
    primes = genPrimes(n)
    0
    #IO.puts("PrimesDone")
    #pfMap = Enum.reduce(2..n, %{}, fn x, pfMap ->
    #  primeFactors(x, Enum.take_while(primes, fn y -> y <= x end), pfMap)
    #end)
    #x = Enum.min_by(Enum.filter(Enum.map(pfMap, fn {x, pf} -> {x, phi(x, pf)} end), fn {x, y} -> Enum.sort(Integer.digits(x)) == Enum.sort(Integer.digits(y)) end), fn {x, y} -> x / y end)
    #elem(x, 0) / elem(x, 1)
  end
end

begin = System.monotonic_time(:nanosecond)
result = Euler70.result
fin = System.monotonic_time(:nanosecond)
IO.puts(to_string(result) <> "\ntook: " <> to_string(fin - begin))
